src/python-bindings/classad2/classad.cpp

```diff
@@ -1 +1,713 @@
+
+static PyObject *
+_classad_size( PyObject *, PyObject * args ) {
+    // _classad_size( self._handle )
+
+    PyObject_Handle * handle = NULL;
+    if(! PyArg_ParseTuple( args, "O", (PyObject **)& handle )) {
+        // PyArg_ParseTuple() has already set an exception for us.
+        return NULL;
+    }
+
+    auto * classAd = (ClassAd *)handle->t;
+    long size = classAd->size();
+    return PyLong_FromLong(size);
+}
+
+
+static PyObject *
+_classad_keys( PyObject *, PyObject * args ) {
+    // _classad_size( self._handle )
+
+    PyObject_Handle * handle = NULL;
+    if(! PyArg_ParseTuple( args, "O", (PyObject **)& handle )) {
+        // PyArg_ParseTuple() has already set an exception for us.
+        return NULL;
+    }
+
+    auto * classAd = (ClassAd *)handle->t;
+
+    PyObject * list = PyList_New(0);
+    if( list == NULL ) {
+        PyErr_SetString( PyExc_MemoryError, "convert_classad_value_to_python" );
+        return NULL;
+    }
+
+    for( auto i = classAd->begin(); i != classAd->end(); ++i ) {
+        PyObject * py_str = PyUnicode_FromString(i->first.c_str());
+        auto rv = PyList_Append( list, py_str );
+        Py_DecRef(py_str);
+
+        if(rv != 0) {
+            // PyList_Append() has already set an exception for us.
+            return NULL;
+        }
+    }
+
+    return list;
+}
+
+
+bool
+isOldAd( const char * in_string ) {
+    while( in_string[0] != '\0' ) {
+        // A leading / starts a comment.
+        if( in_string[0] == '/' || in_string[0] == '[' ) { return false; }
+        if(! isspace(in_string[0])) { return true; }
+        ++in_string;
+    }
+    return false;
+}
+
+
+static PyObject *
+_classad_parse_next( PyObject *, PyObject * args ) {
+    // _classad_parse_next(input_string, int(parser))
+
+    long parser_type = -1;
+    const char * from_string = NULL;
+    if(! PyArg_ParseTuple( args, "zl", & from_string, & parser_type )) {
+        // PyArg_ParseTuple() has already set an exception for us.
+        return NULL;
+    }
+
+
+    // The 'auto' type from the ClassAdFileParseType and the 'auto' type
+    // from version 1 both have weird, undocumented semantics, and they
+    // disagree.  The 'auto' from version 1 only distinguished between
+    // new and old ClassAd serializations, but supported a stream of new-form
+    // ad serializations without requiring that they be a new ClassAd list,
+    // must less one which had a newline after its opening left brace.
+    //
+    // Instead, we add a Python-only parse type, the default, which acts
+    // like version 1, and otherwise passes the other C++ types through.
+    ClassAdFileParseType::ParseType pType = (ClassAdFileParseType::ParseType)parser_type;
+    if( (long)pType == (long)-1 ) {
+        pType = isOldAd(from_string) ? ClassAdFileParseType::Parse_long : ClassAdFileParseType::Parse_new;
+    }
+
+#if defined(WINDOWS)
+    PyErr_SetString(PyExc_NotImplementedError, "Windows doesn't have fmemopen().");
+    return NULL;
+#else
```

Comment:
Is not supporting this in Windows acceptable?

Reply:
We'll have to fix it at some point, but quite frankly, this whole set of functions makes me sad, and I'm not sure it makes sense to fix _just_ the problem with Windows, or to make more serious API and/or implementation changes that result in not having to fix the problem.

Added to HTCONDOR-2163.